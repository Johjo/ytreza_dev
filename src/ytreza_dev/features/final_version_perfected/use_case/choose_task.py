from ytreza_dev.features.final_version_perfected.port.task_repository import TaskRepositoryPort
from ytreza_dev.features.final_version_perfected.types import TaskBase, TaskNew


class ChooseTaskUseCase:
    def __init__(self, task_repository: TaskRepositoryPort):
        self._task_repository = task_repository

    def execute(self, url: str) -> None:
        all_tasks_before = self._task_repository.all_tasks()
        all_tasks_after : list[TaskBase] = []

        all_tasks_after.append(all_tasks_before[0].to_next())

        first_new_index = self._first_new_index(all_tasks_before[1:]) + 1
        i = 1

        while len(all_tasks_before) > i:
            if url == all_tasks_before[i].url:
                all_tasks_after.append(all_tasks_before[i].to_next())
            else:
                if i != first_new_index:
                    all_tasks_after.append(all_tasks_before[i])
                else:
                    all_tasks_after.append(all_tasks_before[i].to_later())
            i += 1

        self._task_repository.save(all_tasks_after)

    @staticmethod
    def _first_new_index(all_tasks_before: list[TaskBase]) -> int:
        for index, task in enumerate(all_tasks_before):
            if isinstance(task, TaskNew):
                return index

        raise NotImplementedError()
