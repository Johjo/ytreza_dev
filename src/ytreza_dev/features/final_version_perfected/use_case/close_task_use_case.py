from ytreza_dev.features.final_version_perfected.port.external_todolist import ExternalTodolistPort
from ytreza_dev.features.final_version_perfected.port.task_repository import TaskFvpRepositoryPort
from ytreza_dev.features.final_version_perfected.types import TaskNever, TaskBase


class CloseTaskUseCase:
    def __init__(self, task_repository: TaskFvpRepositoryPort, external_todolist: ExternalTodolistPort) -> None:
        self._external_todolist = external_todolist
        self._task_repository = task_repository

    def execute(self, url: str) -> None:
        after, closed_task, before = self._extract_before_and_task_and_after(self._task_repository.all_tasks(), url)
        after += self._update_task_following_closed_task(before)
        if all(isinstance(task, TaskNever) for task in after):
            after = [task.to_new() for task in after]
        self._task_repository.save(after)
        self._external_todolist.close_task(url=url, task_id=closed_task.id)

    @staticmethod
    def _update_task_following_closed_task(before: list[TaskBase]) -> list[TaskBase]:
        after : list[TaskBase] = []
        for task in before:
            if isinstance(task, TaskNever):
                after.append(task)
            else:
                after.append(task.to_new())
        return after

    @staticmethod
    def _extract_before_and_task_and_after(before: list[TaskBase], url: str) -> tuple[list[TaskBase], TaskBase,list[TaskBase]]:
        after = []
        # on garde toutes les tâches avant la tâche à fermer
        i = 0
        while before[i].url != url:
            after.append(before[i])
            i += 1

        closed_task = before[i]
        # on ignore la tâche fermée
        before = before[i + 1:]
        return after, closed_task, before
