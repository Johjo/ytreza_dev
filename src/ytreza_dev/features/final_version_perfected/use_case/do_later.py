from ytreza_dev.features.final_version_perfected.port.task_repository import TaskFvpRepositoryPort
from ytreza_dev.features.final_version_perfected.types import TaskBase


class DoLater:
    def __init__(self, task_repository: TaskFvpRepositoryPort) -> None:
        self._task_repository = task_repository

    def execute(self, updated_key: str) -> None:
        before : list[TaskBase] = self._task_repository.all_tasks()

        self._task_repository.save([before[0].to_next()] + [self._update_task(task, updated_key) for task in before[1:]])

    @staticmethod
    def _update_task(task: TaskBase, updated_key: str) -> TaskBase:
        if task.id != updated_key:
            return task

        return task.to_later()
