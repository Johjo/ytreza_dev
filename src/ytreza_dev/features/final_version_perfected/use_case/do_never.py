from ytreza_dev.features.final_version_perfected.types import TaskBase


class DoNever:
    def __init__(self, task_repository):
        self._task_repository = task_repository

    def execute(self, url: str):
        before : list[TaskBase] = self._task_repository.all_tasks()

        self._task_repository.save([before[0].to_next()] + [self._update_task(task, url) for task in before[1:]])

    @staticmethod
    def _update_task(task: TaskBase, url):
        if task.url != url:
            return task

        return task.to_never()
