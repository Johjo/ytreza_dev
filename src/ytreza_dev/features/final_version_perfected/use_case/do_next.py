from ytreza_dev.features.final_version_perfected.port.task_repository import TaskRepositoryPort
from ytreza_dev.features.final_version_perfected.types import TaskBase, TaskNever


class DoNext:
    def __init__(self, task_repository: TaskRepositoryPort):
        self._task_repository = task_repository

    def execute(self, url: str) -> None:
        before: list[TaskBase] = self._task_repository.all_tasks()

        self._task_repository.save([self._update_first_task(before)] + [self.update_task(task, url) for task in before[1:]])

    @staticmethod
    def _update_first_task(before: list[TaskBase]) -> TaskBase:
        if isinstance(before[0], TaskNever):
            return before[0]
        return before[0].to_next()

    @staticmethod
    def update_task(task: TaskBase, url: str) -> TaskBase:
        if task.url != url:
            return task
        return task.to_next()
