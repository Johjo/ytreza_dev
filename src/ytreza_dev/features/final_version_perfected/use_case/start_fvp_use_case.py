from expression import Nothing

from ytreza_dev.features.final_version_perfected.port.task_information_repository import TaskInformationRepositoryPort, \
    TaskInformation
from ytreza_dev.features.final_version_perfected.port.task_repository import FvpRepositoryPort
from ytreza_dev.features.final_version_perfected.port.todolist_reader import TodolistReaderPort
from ytreza_dev.features.final_version_perfected.types import TaskNew, ExternalTask, ExternalProject, Project


class StartFvpUseCase:
    def __init__(self, todolist_reader: TodolistReaderPort, fvp_repository: FvpRepositoryPort, task_information_repository: TaskInformationRepositoryPort) -> None:
        self._task_information_repository = task_information_repository
        self._todolist_reader = todolist_reader
        self._fvp_repository = fvp_repository

    def execute(self) -> None:

        all_external_tasks : list[ExternalTask] = self._todolist_reader.all_tasks()
        self._fvp_repository.save([self._to_task_new(task) for task in all_external_tasks])
        self._task_information_repository.save([self._to_task_information(task) for task in all_external_tasks])

    @staticmethod
    def _to_task_information(external_task: ExternalTask) -> TaskInformation:
        return TaskInformation(
            key=external_task.id,
            title=external_task.name,
            project=Project(key=external_task.project.key, name=external_task.project.name),
            url=external_task.url,
            due_date=Nothing
        )

    @staticmethod
    def _to_task_new(task: ExternalTask) -> TaskNew :
        return TaskNew(id=task.id)

    @staticmethod
    def _to_project(project: ExternalProject) -> Project:
        return Project(key=project.key, name=project.name)
