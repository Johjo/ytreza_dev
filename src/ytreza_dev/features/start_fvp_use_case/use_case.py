from abc import ABCMeta, abstractmethod
from dataclasses import dataclass


@dataclass
class Task:
    name: str
    url: str


class TaskRepository(metaclass=ABCMeta):
    @abstractmethod
    def save(self, tasks: list[Task]) -> None:
        pass


class TodolistReader(metaclass=ABCMeta):
    @abstractmethod
    def all_tasks(self) -> list[Task]:
        pass


class StartFvpUseCase:
    def __init__(self, todolist_reader: TodolistReader, task_repository: TaskRepository) -> None:
        self._todolist_reader = todolist_reader
        self._task_repository = task_repository

    def execute(self) -> None:
        self._task_repository.save(self._todolist_reader.all_tasks())
