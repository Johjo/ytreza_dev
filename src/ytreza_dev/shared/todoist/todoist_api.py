import json
from dataclasses import dataclass
from typing import Any

import requests

@dataclass
class TodoistProject:
    id: str
    name: str


@dataclass
class TodoistTask:
    name: str
    url: str
    id: str
    project: TodoistProject


class TodoistAPI:
    def __init__(self, api_token: str):
        self._api_token = api_token
        self._base_url = "https://api.todoist.com/rest/v2"

    def get_all_tasks(self) -> list[TodoistTask]:
        tasks = json.loads(self._get_all_tasks_json())
        projects : dict[str, Any] = {p["id"]: p for p in json.loads(self._get_all_projects_json())}
        return [self._to_todoist_task(task, projects) for task in tasks]

    def _get_all_tasks_json(self):
        headers = {
            "Authorization": f"Bearer {self._api_token}"
        }
        response = requests.get(f"{self._base_url}/tasks", headers=headers)
        response.raise_for_status()
        json_content = response.text
        return json_content

    def _get_all_projects_json(self):
        headers = {
            "Authorization": f"Bearer {self._api_token}"
        }
        response = requests.get(f"{self._base_url}/projects", headers=headers)
        response.raise_for_status()
        json_content = response.text
        return json_content

    @staticmethod
    def _to_todoist_task(task, projects: [dict[str, Any]]):
        return TodoistTask(
            name=task["content"],
            url=task["url"],
            id=task["id"],
            project=TodoistProject(id=task["project_id"], name=projects[task["project_id"]]["name"])
        )

    def open_task(self, content: str) -> str:
        headers = {
            "Authorization": f"Bearer {self._api_token}",
            "Content-Type": "application/json"
        }
        response = requests.post(f"{self._base_url}/tasks", headers=headers, data=json.dumps({"content": content}))
        response.raise_for_status()

        json_content = response.text
        task = json.loads(json_content)
        return str(task["id"])

    def close_task(self, task_id: str) -> None:
        headers = {
            "Authorization": f"Bearer {self._api_token}",
            "Content-Type": "application/json"
        }
        response = requests.post(f"{self._base_url}/tasks/{task_id}/close", headers=headers)
        response.raise_for_status()

    def task_by_id(self, task_id: str) -> TodoistTask:
        task = json.loads(self._json_task_by_id(task_id))
        projects: dict[str, Any] = {p["id"]: p for p in json.loads(self._get_all_projects_json())}
        return self._to_todoist_task(task, projects)

    def _json_task_by_id(self, task_id: str) -> str:
        headers = {
            "Authorization": f"Bearer {self._api_token}"
        }
        response = requests.get(f"{self._base_url}/tasks/{task_id}", headers=headers)
        response.raise_for_status()
        json_content = response.text
        return json_content
