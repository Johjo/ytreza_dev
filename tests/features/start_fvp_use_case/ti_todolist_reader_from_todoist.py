import os

from approvaltests import verify
from dotenv import load_dotenv

from ytreza_dev.features.start_fvp_use_case.use_case import TodolistReader, Task
from ytreza_dev.shared.todoist.todoist_api import TodoistAPI


class TodolistReaderFromTodoist(TodolistReader):
    def all_tasks(self) -> list[Task]:
        todoist = TodoistAPI(os.getenv("TODOIST_API_TOKEN"))
        tasks = todoist.get_all_tasks()
        return [Task(name=task.name, url=task.url) for task in tasks]


def test_all_tasks() -> None:
    load_dotenv(dotenv_path=".env.test")
    api_token = os.getenv("TODOIST_API_TOKEN")
    assert api_token, "TODOIST_API_TOKEN not found in .env.test"

    todoist_api = TodolistReaderFromTodoist()
    tasks = todoist_api.all_tasks()

    verify("\n".join([str(task) for task in tasks]))
