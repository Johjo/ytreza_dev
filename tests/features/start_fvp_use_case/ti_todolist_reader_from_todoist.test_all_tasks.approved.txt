Task(name=b'T\xc3\xa9l\xc3\xa9chargez des [applications et plugins gratuits suppl\xc3\xa9mentaires](https://todoist.com/fr/downloads?utm_source=todoist&utm_medium=in_app&utm_campaign=onboarding_project&utm_content=inbox) pour \xf0\x9f\x92\xbb,\xe2\x8c\x9a\xef\xb8\x8f,\xf0\x9f\x96\xa5,\xf0\x9f\x93\xb1 et \xf0\x9f\x93\xa7', url='https://app.todoist.com/app/task/9388991121')
Task(name=b'Connecter mon calendrier professionnel \xc3\xa0 Todoist', url='https://app.todoist.com/app/task/9388991491')
Task(name=b'R\xc3\xa9pondez au [questionnaire sur votre m\xc3\xa9thode de productivit\xc3\xa9](https://todoist.com/productivity-methods?utm_source=todoist&utm_medium=in_app&utm_campaign=onboarding_project&utm_content=inbox)', url='https://app.todoist.com/app/task/9388991122')
Task(name=b'Passer en revue ma journ\xc3\xa9e et planifier', url='https://app.todoist.com/app/task/9388991492')
Task(name=b'Ajouter mes routines _professionnelles_', url='https://app.todoist.com/app/task/9388991493')
Task(name=b'[Faire une pause r\xc3\xa9guli\xc3\xa8re](https://blog.doist.com/end-work-day/?utm_source=todoist&utm_medium=in_app&utm_campaign=onboarding_project&utm_content=work) \xf0\x9f\x8c\x96', url='https://app.todoist.com/app/task/9388991494')
Task(name=b'[Les int\xc3\xa9grations de Todoist qui boosteront votre productivit\xc3\xa9](https://blog.doist.com/todoist-integrations-productivity/?utm_source=todoist&utm_medium=in_app&utm_campaign=onboarding_project&utm_content=work)', url='https://app.todoist.com/app/task/9388991495')
Task(name=b'Explorer les [mod\xc3\xa8les professionnels de Todoist](https://todoist.com/fr/templates/category/work?utm_source=todoist&utm_medium=in_app&utm_campaign=onboarding_project&utm_content=work)', url='https://app.todoist.com/app/task/9388991496')
